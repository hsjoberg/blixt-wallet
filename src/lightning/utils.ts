import deepMapKeys from "deep-map-keys";

export const fixGrpcJsonResponse = <T>(response: { [keys: string]: any })  =>
  deepMapKeys<T>(response, (key) => {
    if (key[key.length - 1] === "_") {
      return key.slice(0, -1);
    }
    return "$" + key;
});
