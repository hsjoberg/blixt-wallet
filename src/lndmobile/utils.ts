import { NativeModules } from "react-native";
import deepMapKeys from "deep-map-keys";
import * as base64 from "base64-js";
import * as $protobuf from "protobufjs";
import { lnrpc } from "../../proto/proto";

const { LndMobile } = NativeModules;

export interface ISendRequestClass<IReq, Req> {
  create: (options: IReq) => Req;
  encode: (request: Req) => $protobuf.Writer;
}

export interface ISendResponseClass<Res> {
  decode: (reader: $protobuf.Reader | Uint8Array) => Res;
  toObject(message: Res, options?: $protobuf.IConversionOptions): { [k: string]: any };
}

export interface ISyncCommandOptions<IReq, Req, Res> {
    request: ISendRequestClass<IReq, Req>;
    response: ISendResponseClass<Res>;
    method: string;
    options: IReq;
}

export interface IStreamCommandOptions<IReq, Req> {
    request: ISendRequestClass<IReq, Req>;
    method: string;
    options: IReq;
}

export interface IStreamResultOptions<Res> {
  base64Result: string;
  response: ISendResponseClass<Res>;
}

/**
 * @throws
 */
export const sendCommand = async <IReq, Req, Res>({ request, response, method, options }: ISyncCommandOptions<IReq, Req, Res>): Promise<Res> => {
  try {
    const instance = request.create(options);
    const b64 = await LndMobile.sendCommand(method, base64.fromByteArray(request.encode(instance).finish()));
    return response.decode(base64.toByteArray(b64.data));
  } catch (e) { throw e; }
};

export const sendStreamCommand = async <IReq, Req>({ request, method, options }: IStreamCommandOptions<IReq, Req>, streamOnlyOnce: boolean = false): Promise<string> =>  {
  const instance = request.create(options);
  const response = await LndMobile.sendStreamCommand(method, base64.fromByteArray(request.encode(instance).finish()), streamOnlyOnce);
  return response;
};

export const decodeStreamResult = <Res>( { base64Result, response }: IStreamResultOptions<Res>): Res => {
  return response.decode(base64.toByteArray(base64Result));
};
