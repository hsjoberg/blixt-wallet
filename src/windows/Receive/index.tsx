import React from "react"
import { createStackNavigator } from "@react-navigation/stack";

import ReceiveSetup from "./ReceiveSetup";
import ReceiveQr from "./ReceiveQr";
import { lnrpc } from "../../../proto/proto";

const Stack = createStackNavigator();

export type ReceiveStackParamList = {
  ReceiveSetup: undefined;
  ReceiveQr: {
    invoice: lnrpc.AddInvoiceResponse;
  };
}

export default () => {
  return (
    <Stack.Navigator initialRouteName="ReceiveSetup" screenOptions={{ headerShown: false, animationEnabled: false }}>
      <Stack.Screen name="ReceiveSetup" component={ReceiveSetup} />
      <Stack.Screen name="ReceiveQr" component={ReceiveQr} />
    </Stack.Navigator>
  )
}