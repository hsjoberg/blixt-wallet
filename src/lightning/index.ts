import { NativeModules } from "react-native";
import { fixGrpcJsonResponse } from "./utils";
import { ITransaction } from "../storage/database/transaction";
export * from "./channel";
const { LndGrpc } = NativeModules;

/**
 * @throws
 */
export const unlockWallet = async (password: string): Promise<any> => {
  try {
    const responseString = await LndGrpc.unlockWallet(password);
    return fixGrpcJsonResponse(JSON.parse(responseString));
  } catch (e) { throw JSON.parse(e.message); }
};

export interface IGetInfoResponse {
  pubkey: string;
}

/**
 * @throws
 */
export const getInfo = async (): Promise<IGetInfoResponse> => {
  try {
    const responseString = await LndGrpc.getInfo();
    const response = JSON.parse(responseString);
    return {
      pubkey: response.identityPubkey_,
    };
  } catch (e) { throw JSON.parse(e.message); }
};

export interface IPaymentRouteHops {
  amtToForwardMsat: number;
  amtToForward: number;
  chanCapacity: number;
  chanId: number;
  expiry: number;
  feeMsat: number;
  fee: number;
  pubkey: string;
}

export interface ISendPaymentSyncResponse {
  paymentError: string;
  paymentHash: {
    bytes: number[];
    hash: number;
  };
  paymentPreImage: {
    bytes: number[];
    hash: number;
  };
  paymentRoute: {
    bitField0: number;
    paymentRoute: IPaymentRouteHops[]
  };
}

/**
 * @throws
 */
export const sendPaymentSync = async (paymentRequest: string): Promise<ISendPaymentSyncResponse> => {
  try {
    const responseString = await LndGrpc.sendPaymentSync(paymentRequest);
    const response = fixGrpcJsonResponse<ISendPaymentSyncResponse>(JSON.parse(responseString));
    return response;
  } catch (e) { throw JSON.parse(e.message); }
};


export interface IAddInvoiceResponse {
  paymentRequest: string;
  addIndex: number;
  rHash: any; // TODO
}
/**
 * @throws
 */
export const addInvoice = async (sat: number, memo: string, expiry: number = 3600): Promise<IAddInvoiceResponse> => {
  try {
    const responseString = await LndGrpc.addInvoice(sat, memo, expiry);
    const response = fixGrpcJsonResponse<IAddInvoiceResponse>(JSON.parse(responseString));
    return response;
  } catch (e) { throw JSON.parse(e.message); }
};

/**
 * @throws
 */
export const lookupInvoice = async (rHash: string): Promise<any> => {
  try {
    const responseString = await LndGrpc.lookupInvoice(rHash);
    const response = fixGrpcJsonResponse<any>(JSON.parse(responseString));
    return response;
  } catch (e) { throw JSON.parse(e.message); }
};


export interface IDecodePayReqResponse {
  cltvExpiry: number; // Presumably blocks
  descriptionHash: string;
  description: string;
  destination: string;
  expiry: number; // actual seconds (NOT a timestamp)
  fallbackAddr: string;
  numSatoshis: number;
  paymentHash: string;
  routeHints: any[]; // TODO
  timestamp: number;
}
/**
 * @throws
 */
export const decodePayReq = async (bolt11: string): Promise<IDecodePayReqResponse> => {
  try {
    const responseString = await LndGrpc.decodePayReq(bolt11);
    const response = fixGrpcJsonResponse<IDecodePayReqResponse>(JSON.parse(responseString));
    return response;
  } catch (e) { throw JSON.parse(e.message); }
};
