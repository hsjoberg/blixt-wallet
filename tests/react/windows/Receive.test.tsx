import React from "react";
import { act, render, fireEvent, waitFor, within } from "@testing-library/react-native";
import { StoreProvider } from "easy-peasy";
import Long from "long";
import * as base64 from "base64-js";

import Overview from "../../../src/windows/Overview";
import Receive from "../../../src/windows/Receive";
import { createNavigationContainer, setupStore, setDefaultAsyncStorage, initCommonStore } from "../../utils";
import { createStackNavigator } from "@react-navigation/stack";
import { NavigationContainer } from "@react-navigation/native";
import { DeviceEventEmitter } from "react-native";
import { lnrpc } from "../../../proto/lightning";
import { hexToUint8Array } from "../../../src/utils";

jest.setTimeout(10000);

const AppContainer = createNavigationContainer(Receive, "Receive");

it("renders correctly", () => {
  const store = setupStore();

  const { unmount, toJSON } = render(
    <StoreProvider store={store}>
      {AppContainer}
    </StoreProvider>
  );
  expect(toJSON()).toMatchSnapshot();

  unmount();
});

it("is possible to create an invoice and go to the QR screen", async () => {
  await setDefaultAsyncStorage();
  const store = await initCommonStore(true);

  const { queryByTestId, unmount } = render(
    <StoreProvider store={store}>
      {AppContainer}
    </StoreProvider>
  );

  const inputAmountSat = queryByTestId("input-amount-sat");
  const inputMessage = queryByTestId("input-message");
  const createInvoiceButton = queryByTestId("create-invoice");

  expect(createInvoiceButton).not.toBeNull();
  expect(inputAmountSat).not.toBeNull();
  expect(inputMessage).not.toBeNull();

  act(() => void fireEvent.changeText(inputAmountSat!, "10000"));
  expect(inputAmountSat!.props.value).toBe("10000");
  act(() => void fireEvent.changeText(inputMessage!, "A test invoice"));
  expect(inputMessage!.props.value).toBe("A test invoice");

  // Pressing createInvoiceButton moves us
  // to the next window, from ReceiveSetup.tsx to ReceiveQr.tsx
  act(() => void fireEvent.press(createInvoiceButton!));

  await waitFor(() => expect(store.getState().transaction.transactions).toHaveLength(1));
  const paymentRequestString = await waitFor(() => queryByTestId("payment-request-string"));
  const expireString = await waitFor(() => queryByTestId("expire"));
  const payAmountString = await waitFor(() => queryByTestId("pay-amount"));

  expect(paymentRequestString).not.toBeNull();
  expect(expireString!.children[0]!.children!.join()).toContain("Expires in");
  expect(payAmountString!.children.join()).toContain("10 000");

  unmount();
  store.getActions().receive.deinitialize();
});

test("invoice appears on the transaction list", async () => {
  const store = await initCommonStore(true);
  const RootStack = createStackNavigator();

  const Container = (
    <NavigationContainer>
      <RootStack.Navigator>
        <RootStack.Screen name="Overview" component={Overview} />
        <RootStack.Screen name="Receive" component={Receive} />
      </RootStack.Navigator>
    </NavigationContainer>
  );

  const { getByTestId, findByTestId, unmount, getAllByTestId } = render(
    <StoreProvider store={store}>
      {Container}
    </StoreProvider>
  );

  // See if we can find the Overview transaction list
  const txList = getByTestId("TX_LIST");
  expect(Array.isArray(txList.children)).toBe(true);

  // Navigate to Receive screen
  const footerReceiveButton = await findByTestId("FOOTER_RECEIVE");
  fireEvent.press(footerReceiveButton);

  // Create an invoice
  const inputAmountSat = await findByTestId("input-amount-sat");
  const inputMessage = getByTestId("input-message");
  const createInvoiceButton = getByTestId("create-invoice");
  act(() => void fireEvent.changeText(inputAmountSat!, "10000"));
  act(() => void fireEvent.changeText(inputMessage!, "A test invoice"));
  act(() => void fireEvent.press(createInvoiceButton!));

  // Expect an invoice to be created
  await waitFor(() => expect(store.getState().transaction.transactions).toHaveLength(1));

  // Go back to the Overview screen
  const goBackButton = getByTestId("header-back");
  fireEvent.press(goBackButton);

  // Expect a new invoice to be created with state OPEN
  const tx = store.getState().transaction.transactions[0];
  expect(tx.status).toBe("OPEN");
  expect(within(txList).queryAllByText("Open")).not.toBeNull();

  // Pay invoice
  const invoice = lnrpc.Invoice.create({
    amtPaid: Long.fromValue(10000),
    amtPaidSat: Long.fromValue(10000),
    amtPaidMsat: Long.fromValue(1000000),
    rHash: hexToUint8Array(tx.rHash),
    // paymentRequest: "payreq",
    expiry: Long.fromValue(40),
    isKeysend: false,
    state: lnrpc.Invoice.InvoiceState.SETTLED,
  })
  act(() => {
    DeviceEventEmitter.emit("SubscribeInvoices", {
      data: base64.fromByteArray(lnrpc.Invoice.encode(invoice).finish())
    });
  });

  // Expect the invoice to have status SETTLED
  await waitFor(async () => {
    expect(store.getState().transaction.transactions[0].status).toBe("SETTLED");
  });

  expect(within(txList).queryAllByText("Open")).toEqual(expect.arrayContaining([]));

  unmount();
});
