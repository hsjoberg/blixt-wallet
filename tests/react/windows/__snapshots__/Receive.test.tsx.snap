// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  collapsable={true}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <RNCSafeAreaProvider
      style={
        Array [
          Object {
            "flex": 1,
          },
          undefined,
        ]
      }
    >
      <View
        pointerEvents="box-none"
        style={
          Array [
            Object {
              "zIndex": 1,
            },
            false,
          ]
        }
      >
        <View
          accessibilityElementsHidden={false}
          importantForAccessibility="auto"
          pointerEvents="box-none"
          style={null}
        >
          <View
            pointerEvents="box-none"
            style={
              Object {
                "bottom": 0,
                "left": 0,
                "opacity": 1,
                "position": "absolute",
                "right": 0,
                "top": 0,
                "zIndex": 0,
              }
            }
          >
            <View
              style={
                Object {
                  "backgroundColor": "rgb(255, 255, 255)",
                  "borderBottomColor": "rgb(216, 216, 216)",
                  "flex": 1,
                  "shadowColor": "rgb(216, 216, 216)",
                  "shadowOffset": Object {
                    "height": 0.5,
                    "width": 0,
                  },
                  "shadowOpacity": 0.85,
                  "shadowRadius": 0,
                }
              }
            />
          </View>
          <View
            pointerEvents="box-none"
            style={
              Object {
                "height": 64,
                "maxHeight": undefined,
                "minHeight": undefined,
                "opacity": undefined,
                "transform": undefined,
              }
            }
          >
            <View
              pointerEvents="none"
              style={
                Object {
                  "height": 20,
                }
              }
            />
            <View
              pointerEvents="box-none"
              style={
                Object {
                  "alignItems": "center",
                  "flex": 1,
                  "flexDirection": "row",
                  "justifyContent": "center",
                }
              }
            >
              <View
                pointerEvents="box-none"
                style={
                  Object {
                    "marginHorizontal": 16,
                    "opacity": 1,
                  }
                }
              >
                <Text
                  accessibilityRole="header"
                  numberOfLines={1}
                  style={
                    Object {
                      "color": "rgb(28, 28, 30)",
                      "fontSize": 17,
                      "fontWeight": "600",
                    }
                  }
                >
                  a
                </Text>
              </View>
            </View>
          </View>
        </View>
      </View>
      <View
        style={
          Object {
            "flex": 1,
          }
        }
      >
        <View
          pointerEvents="box-none"
          style={
            Object {
              "bottom": 0,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            }
          }
        >
          <View
            collapsable={false}
            style={
              Object {
                "opacity": 1,
              }
            }
          />
          <View
            accessibilityElementsHidden={false}
            closing={false}
            gestureVelocityImpact={0.3}
            importantForAccessibility="auto"
            pointerEvents="box-none"
            style={
              Object {
                "bottom": 0,
                "left": 0,
                "position": "absolute",
                "right": 0,
                "top": 0,
              }
            }
            transitionSpec={
              Object {
                "close": Object {
                  "animation": "spring",
                  "config": Object {
                    "damping": 500,
                    "mass": 3,
                    "overshootClamping": true,
                    "restDisplacementThreshold": 10,
                    "restSpeedThreshold": 10,
                    "stiffness": 1000,
                  },
                },
                "open": Object {
                  "animation": "spring",
                  "config": Object {
                    "damping": 500,
                    "mass": 3,
                    "overshootClamping": true,
                    "restDisplacementThreshold": 10,
                    "restSpeedThreshold": 10,
                    "stiffness": 1000,
                  },
                },
              }
            }
          >
            <View
              pointerEvents="box-none"
              style={
                Object {
                  "flex": 1,
                }
              }
            >
              <View
                collapsable={false}
                style={
                  Object {
                    "flex": 1,
                    "transform": Array [
                      Object {
                        "translateX": 0,
                      },
                      Object {
                        "translateX": 0,
                      },
                    ],
                  }
                }
              >
                <View
                  pointerEvents="none"
                  style={
                    Object {
                      "backgroundColor": "rgb(242, 242, 242)",
                      "bottom": 0,
                      "left": 0,
                      "position": "absolute",
                      "shadowColor": "#000",
                      "shadowOffset": Object {
                        "height": 1,
                        "width": -1,
                      },
                      "shadowOpacity": 0.3,
                      "shadowRadius": 5,
                      "top": 0,
                      "width": 3,
                    }
                  }
                />
                <View
                  style={
                    Array [
                      Object {
                        "flex": 1,
                        "overflow": "hidden",
                      },
                      Array [
                        Object {
                          "backgroundColor": "rgb(242, 242, 242)",
                        },
                        undefined,
                      ],
                    ]
                  }
                >
                  <View
                    style={
                      Object {
                        "flex": 1,
                        "flexDirection": "column-reverse",
                      }
                    }
                  >
                    <View
                      style={
                        Object {
                          "flex": 1,
                        }
                      }
                    >
                      <View
                        style={
                          Object {
                            "flex": 1,
                          }
                        }
                      >
                        <View
                          pointerEvents="box-none"
                          style={
                            Array [
                              Object {
                                "zIndex": 1,
                              },
                              false,
                            ]
                          }
                        >
                          <View
                            accessibilityElementsHidden={false}
                            importantForAccessibility="auto"
                            pointerEvents="box-none"
                            style={null}
                          >
                            <View
                              pointerEvents="box-none"
                              style={
                                Object {
                                  "bottom": 0,
                                  "left": 0,
                                  "position": "absolute",
                                  "right": 0,
                                  "top": 0,
                                  "transform": Array [
                                    Object {
                                      "translateX": 0,
                                    },
                                  ],
                                  "zIndex": 0,
                                }
                              }
                            >
                              <View
                                style={
                                  Object {
                                    "backgroundColor": "#bc6610",
                                    "borderBottomColor": "rgb(216, 216, 216)",
                                    "flex": 1,
                                    "shadowColor": "rgb(216, 216, 216)",
                                    "shadowOffset": Object {
                                      "height": 0.5,
                                      "width": 0,
                                    },
                                    "shadowOpacity": 0.85,
                                    "shadowRadius": 0,
                                  }
                                }
                              />
                            </View>
                            <View
                              pointerEvents="box-none"
                              style={
                                Object {
                                  "height": 64,
                                  "maxHeight": undefined,
                                  "minHeight": undefined,
                                  "opacity": undefined,
                                  "transform": undefined,
                                }
                              }
                            >
                              <View
                                pointerEvents="none"
                                style={
                                  Object {
                                    "height": 20,
                                  }
                                }
                              />
                              <View
                                pointerEvents="box-none"
                                style={
                                  Object {
                                    "alignItems": "center",
                                    "flex": 1,
                                    "flexDirection": "row",
                                    "justifyContent": "center",
                                  }
                                }
                              >
                                <View
                                  pointerEvents="box-none"
                                  style={
                                    Object {
                                      "marginHorizontal": 16,
                                      "transform": Array [
                                        Object {
                                          "translateX": 0,
                                        },
                                      ],
                                    }
                                  }
                                >
                                  <Text
                                    accessibilityRole="header"
                                    numberOfLines={1}
                                    style={
                                      Object {
                                        "color": "#e5eaea",
                                        "fontSize": 17,
                                        "fontWeight": "600",
                                      }
                                    }
                                  >
                                    Receive
                                  </Text>
                                </View>
                              </View>
                            </View>
                          </View>
                        </View>
                        <View
                          style={
                            Object {
                              "flex": 1,
                            }
                          }
                        >
                          <View
                            pointerEvents="box-none"
                            style={
                              Object {
                                "bottom": 0,
                                "left": 0,
                                "position": "absolute",
                                "right": 0,
                                "top": 0,
                              }
                            }
                          >
                            <View
                              collapsable={false}
                              style={
                                Object {
                                  "opacity": 1,
                                }
                              }
                            />
                            <View
                              accessibilityElementsHidden={false}
                              closing={false}
                              gestureVelocityImpact={0.3}
                              importantForAccessibility="auto"
                              pointerEvents="box-none"
                              style={
                                Object {
                                  "bottom": 0,
                                  "left": 0,
                                  "position": "absolute",
                                  "right": 0,
                                  "top": 0,
                                }
                              }
                              transitionSpec={
                                Object {
                                  "close": Object {
                                    "animation": "spring",
                                    "config": Object {
                                      "damping": 500,
                                      "mass": 3,
                                      "overshootClamping": true,
                                      "restDisplacementThreshold": 10,
                                      "restSpeedThreshold": 10,
                                      "stiffness": 1000,
                                    },
                                  },
                                  "open": Object {
                                    "animation": "spring",
                                    "config": Object {
                                      "damping": 500,
                                      "mass": 3,
                                      "overshootClamping": true,
                                      "restDisplacementThreshold": 10,
                                      "restSpeedThreshold": 10,
                                      "stiffness": 1000,
                                    },
                                  },
                                }
                              }
                            >
                              <View
                                pointerEvents="box-none"
                                style={
                                  Object {
                                    "flex": 1,
                                  }
                                }
                              >
                                <View
                                  collapsable={false}
                                  style={
                                    Object {
                                      "flex": 1,
                                    }
                                  }
                                >
                                  <View
                                    style={
                                      Array [
                                        Object {
                                          "flex": 1,
                                          "overflow": "hidden",
                                        },
                                        Array [
                                          Object {
                                            "backgroundColor": "rgb(242, 242, 242)",
                                          },
                                          Object {
                                            "backgroundColor": "transparent",
                                          },
                                        ],
                                      ]
                                    }
                                  >
                                    <View
                                      style={
                                        Object {
                                          "flex": 1,
                                          "flexDirection": "column-reverse",
                                        }
                                      }
                                    >
                                      <View
                                        style={
                                          Object {
                                            "flex": 1,
                                          }
                                        }
                                      >
                                        <View
                                          style={
                                            Object {
                                              "backgroundColor": "#fff",
                                              "flex": 1,
                                              "height": 1334,
                                            }
                                          }
                                        >
                                          <View
                                            style={
                                              Array [
                                                Object {},
                                                Object {
                                                  "display": "flex",
                                                  "flex": 1,
                                                  "height": "100%",
                                                  "justifyContent": "space-between",
                                                  "padding": 16,
                                                },
                                                undefined,
                                              ]
                                            }
                                          >
                                            <View
                                              style={
                                                Array [
                                                  Object {},
                                                  Object {},
                                                ]
                                              }
                                            >
                                              <TouchableOpacity
                                                activeOpacity={1}
                                                style={
                                                  Object {
                                                    "alignItems": "center",
                                                    "backgroundColor": "transparent",
                                                    "borderBottomWidth": 1,
                                                    "borderColor": "#D9D5DC",
                                                    "borderLeftWidth": 0,
                                                    "borderRadius": undefined,
                                                    "borderRightWidth": 0,
                                                    "borderTopWidth": 0,
                                                    "borderWidth": undefined,
                                                    "flexDirection": "row",
                                                    "marginLeft": 2,
                                                    "marginTop": 8,
                                                  }
                                                }
                                              >
                                                <Text
                                                  style={
                                                    Array [
                                                      Object {
                                                        "color": "#575757",
                                                        "fontSize": 17,
                                                        "paddingRight": 5,
                                                      },
                                                      Object {
                                                        "fontSize": 17,
                                                        "width": 105,
                                                      },
                                                    ]
                                                  }
                                                >
                                                  Amount ₿
                                                </Text>
                                                <TextInput
                                                  allowFontScaling={true}
                                                  editable={true}
                                                  keyboardType="numeric"
                                                  placeholder="0"
                                                  placeholderTextColor="#575757"
                                                  rejectResponderTermination={true}
                                                  style={
                                                    Object {
                                                      "color": "#000",
                                                      "flex": 1,
                                                      "fontSize": 17,
                                                      "height": 50,
                                                      "paddingLeft": 5,
                                                      "paddingRight": 5,
                                                      "top": 1.5,
                                                    }
                                                  }
                                                  testID="input-amount-sat"
                                                  underlineColorAndroid="rgba(0,0,0,0)"
                                                />
                                              </TouchableOpacity>
                                              <TouchableOpacity
                                                activeOpacity={1}
                                                style={
                                                  Object {
                                                    "alignItems": "center",
                                                    "backgroundColor": "transparent",
                                                    "borderBottomWidth": 1,
                                                    "borderColor": "#D9D5DC",
                                                    "borderLeftWidth": 0,
                                                    "borderRadius": undefined,
                                                    "borderRightWidth": 0,
                                                    "borderTopWidth": 0,
                                                    "borderWidth": undefined,
                                                    "flexDirection": "row",
                                                    "marginLeft": 2,
                                                    "marginTop": 16,
                                                  }
                                                }
                                              >
                                                <Text
                                                  style={
                                                    Array [
                                                      Object {
                                                        "color": "#575757",
                                                        "fontSize": 17,
                                                        "paddingRight": 5,
                                                      },
                                                      Object {
                                                        "fontSize": 17,
                                                        "width": 105,
                                                      },
                                                    ]
                                                  }
                                                >
                                                  Amount USD
                                                </Text>
                                                <TextInput
                                                  allowFontScaling={true}
                                                  editable={true}
                                                  keyboardType="numeric"
                                                  placeholder="0.00"
                                                  placeholderTextColor="#575757"
                                                  rejectResponderTermination={true}
                                                  style={
                                                    Object {
                                                      "color": "#000",
                                                      "flex": 1,
                                                      "fontSize": 17,
                                                      "height": 50,
                                                      "paddingLeft": 5,
                                                      "paddingRight": 5,
                                                      "top": 1.5,
                                                    }
                                                  }
                                                  underlineColorAndroid="rgba(0,0,0,0)"
                                                />
                                              </TouchableOpacity>
                                              <TouchableOpacity
                                                activeOpacity={1}
                                                style={
                                                  Object {
                                                    "alignItems": "center",
                                                    "backgroundColor": "transparent",
                                                    "borderBottomWidth": 1,
                                                    "borderColor": "#D9D5DC",
                                                    "borderLeftWidth": 0,
                                                    "borderRadius": undefined,
                                                    "borderRightWidth": 0,
                                                    "borderTopWidth": 0,
                                                    "borderWidth": undefined,
                                                    "flexDirection": "row",
                                                    "marginLeft": 2,
                                                    "marginTop": 16,
                                                  }
                                                }
                                              >
                                                <Text
                                                  style={
                                                    Array [
                                                      Object {
                                                        "color": "#575757",
                                                        "fontSize": 17,
                                                        "paddingRight": 5,
                                                      },
                                                      Object {
                                                        "fontSize": 17,
                                                        "width": 105,
                                                      },
                                                    ]
                                                  }
                                                >
                                                  Payer
                                                </Text>
                                                <TextInput
                                                  allowFontScaling={true}
                                                  editable={true}
                                                  placeholder="For bookkeeping (optional)"
                                                  placeholderTextColor="#575757"
                                                  rejectResponderTermination={true}
                                                  style={
                                                    Object {
                                                      "color": "#000",
                                                      "flex": 1,
                                                      "fontSize": 17,
                                                      "height": 50,
                                                      "paddingLeft": 5,
                                                      "paddingRight": 5,
                                                      "top": 1.5,
                                                    }
                                                  }
                                                  underlineColorAndroid="rgba(0,0,0,0)"
                                                  value=""
                                                />
                                              </TouchableOpacity>
                                              <TouchableOpacity
                                                activeOpacity={1}
                                                style={
                                                  Object {
                                                    "alignItems": "center",
                                                    "backgroundColor": "transparent",
                                                    "borderBottomWidth": 1,
                                                    "borderColor": "#D9D5DC",
                                                    "borderLeftWidth": 0,
                                                    "borderRadius": undefined,
                                                    "borderRightWidth": 0,
                                                    "borderTopWidth": 0,
                                                    "borderWidth": undefined,
                                                    "flexDirection": "row",
                                                    "marginLeft": 2,
                                                    "marginTop": 16,
                                                  }
                                                }
                                              >
                                                <Text
                                                  style={
                                                    Array [
                                                      Object {
                                                        "color": "#575757",
                                                        "fontSize": 17,
                                                        "paddingRight": 5,
                                                      },
                                                      Object {
                                                        "fontSize": 17,
                                                        "width": 105,
                                                      },
                                                    ]
                                                  }
                                                >
                                                  Message
                                                </Text>
                                                <TextInput
                                                  allowFontScaling={true}
                                                  editable={true}
                                                  placeholder="Message to payer (optional)"
                                                  placeholderTextColor="#575757"
                                                  rejectResponderTermination={true}
                                                  style={
                                                    Object {
                                                      "color": "#000",
                                                      "flex": 1,
                                                      "fontSize": 17,
                                                      "height": 50,
                                                      "paddingLeft": 5,
                                                      "paddingRight": 5,
                                                      "top": 1.5,
                                                    }
                                                  }
                                                  testID="input-message"
                                                  underlineColorAndroid="rgba(0,0,0,0)"
                                                  value=""
                                                />
                                              </TouchableOpacity>
                                            </View>
                                            <View
                                              style={
                                                Array [
                                                  Object {},
                                                  Object {
                                                    "marginBottom": 2,
                                                  },
                                                ]
                                              }
                                            >
                                              <View
                                                style={
                                                  Array [
                                                    Object {},
                                                    Object {
                                                      "marginTop": 0,
                                                    },
                                                  ]
                                                }
                                              >
                                                <TouchableOpacity
                                                  activeOpacity={0.5}
                                                  block={true}
                                                  disabled={true}
                                                  primary={true}
                                                  style={
                                                    Object {
                                                      "alignItems": "center",
                                                      "alignSelf": "stretch",
                                                      "backgroundColor": "#b5b5b5",
                                                      "borderBottomWidth": null,
                                                      "borderColor": "#007aff",
                                                      "borderLeftWidth": null,
                                                      "borderRadius": 5,
                                                      "borderRightWidth": null,
                                                      "borderTopWidth": null,
                                                      "borderWidth": undefined,
                                                      "elevation": 2,
                                                      "flexDirection": "row",
                                                      "height": 45,
                                                      "justifyContent": "center",
                                                      "paddingBottom": 6,
                                                      "paddingTop": 6,
                                                      "shadowColor": undefined,
                                                      "shadowOffset": undefined,
                                                      "shadowOpacity": undefined,
                                                      "shadowRadius": undefined,
                                                    }
                                                  }
                                                  testID="create-invoice"
                                                >
                                                  <ActivityIndicator
                                                    animating={true}
                                                    color="#e5eaea"
                                                    hidesWhenStopped={true}
                                                    size="large"
                                                    style={
                                                      Object {
                                                        "height": 80,
                                                      }
                                                    }
                                                  />
                                                </TouchableOpacity>
                                              </View>
                                            </View>
                                          </View>
                                        </View>
                                      </View>
                                    </View>
                                  </View>
                                </View>
                              </View>
                            </View>
                          </View>
                        </View>
                      </View>
                    </View>
                  </View>
                </View>
              </View>
            </View>
          </View>
        </View>
      </View>
    </RNCSafeAreaProvider>
  </View>
</View>
`;
